tool/automation/ImageValidation.kts

```diff
@@ -0,0 +1,221 @@
+ * WARNING: the function depends on the assumption that tag pattern ...
+ * ... is "<year>.<buld number>-<OS>".
+ */
+fun getPrevDockerImageId(imageId: String): String {
+    var curImageTag = imageId.split(":")[1]
+    var curImageTagElems = curImageTag.split(".")
+
+    if (curImageTagElems.size < 2) {
+        // image is highly likely doesn't correspond to pattern
+        throw IllegalArgumentException("Unable to auto-determine previous image tag - it doesn't correspond to pattern: $imageId")
+    }
+
+    // handling 2 types: 2022.04-OS and 2022.04.2-OS
+    val isMinorRelease = curImageTagElems.size > 2
+
+    var imageBuildNum = if (isMinorRelease) curImageTagElems[2].split("-")[0]
+                            else curImageTagElems[1].split("-")[0]
+
+    var oldBuildNumber = Integer.parseInt(imageBuildNum) - 1
+
+    // -- construct old image tag based on retrieved information from the current one
+    // -- -- adding "0" since build number has at least 2 digits
+    val oldBuildNumString = if (oldBuildNumber < 10 && !isMinorRelease) ("0" + oldBuildNumber)
+                                else oldBuildNumber
+
+    // Replace current image's numberic part of tag with determined "old" value, e.g. "2022.04.2-" -> "2022.04.1-"
+    val originalImageTagPart = if (isMinorRelease) (curImageTagElems[0] + "." + curImageTagElems[1] + "." + imageBuildNum + "-")
+                                else (curImageTagElems[0] + "." + imageBuildNum + "-")
+    val determinedOldImageTagPart = if (isMinorRelease)  (curImageTagElems[0] + "." + curImageTagElems[1] + "." + oldBuildNumString + "-")
+                            else (curImageTagElems[0] + "." + oldBuildNumString + "-")
+
+    val oldImageId = imageId.replace(originalImageTagPart, determinedOldImageTagPart)
+    return oldImageId
+}
+
+
+/**
+ * Tries to pull Docker image from registry.
+ * @param name - docker image fully-qualified domain name
+ * @return true if image had been successfully pulled, false otherwise
+ */
+fun pullDockerImage(name: String): Boolean {
+    val cmdResult = this.executeCommand("docker pull $name", true) ?: ""
+
+    // using success messages since some errors from docker daemon (e.g. invalid platform type) are not ...
+    // ... captured by Kotlin's ProcessBuilder.
+    val successMessages = arrayOf("Pull complete", "Image is up to date")
+    return successMessages.any { cmdResult.contains(it, ignoreCase = true) }
+}
+
+
+/**
+ * Validates Docker image size.
+ * Criteria: it shouldn't increase by more than threshold (@see ALLOWED_IMAGE_SIZE_INCREASE_THRESHOLD_PERCENT).
+ * @param currentName - name of original Docker image
+ * @param previousName - name of previous Docker image
+ * @return true if image size increase suppressed given threshold; false otherwise (including situation when ...
+ * ... it wasn't possible to determine any of image sizes)
+ */
+fun imageSizeIncreasedTooMuch(currentName: String, previousName: String): Boolean {
```

Comment:
What about a decreasing image size? Should we also catch such situations?

Reply:
Great idea - I think that'd be reasonable.
An unexpected removal or certain layer might indicate the change within one of our dependencies, for example: base image, utility tools, etc. Even if we wouldn't observe error during image creation and its bootstrap (entrypoint execution, start up of services), It **might** result into inability to run a particular script in runtime.
I will change the function to check percentage difference by modulo, not an absolute value.